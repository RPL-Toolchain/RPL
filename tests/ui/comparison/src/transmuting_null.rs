//@revisions: inline normal
//@[inline] compile-flags: -Zinline-mir=false
#[cfg_attr(test, test)]
fn base_case() {
    unsafe {
        let x: &u64 = std::mem::transmute(0 as *const u64);
        //~^ transmuting_null
        dbg!(x);

        let x: &u64 = std::mem::transmute(std::ptr::null::<u64>());
        //~^ transmuting_null
        dbg!(x);
    }
}

#[cfg_attr(test, test)]
fn cross_function_null_ptr() {
    const fn const_null_ptr<T>() -> *const T {
        std::ptr::null()
    }
    fn null_ptr<T>() -> *const T {
        std::ptr::null()
    }
    unsafe {
        let x: &u64 = std::mem::transmute(const_null_ptr::<u64>());
        //~^ transmuting_null
        dbg!(x);

        let x: &u64 = std::mem::transmute(null_ptr::<u64>());
        //~^ transmuting_null
        dbg!(x);
    }
}

#[cfg_attr(test, test)]
fn cross_statement() {
    let null_ptr = 0 as *const u64;
    unsafe {
        let x: &u64 = std::mem::transmute(null_ptr);
        //~^ transmuting_null
        dbg!(x);
    }

    let null_ptr = std::ptr::null::<u64>();
    unsafe {
        let x: &u64 = std::mem::transmute(null_ptr);
        //~^ transmuting_null
        dbg!(x);
    }
}

pub(crate) fn main() {
    base_case();
    cross_function_null_ptr();
    cross_statement();
}
