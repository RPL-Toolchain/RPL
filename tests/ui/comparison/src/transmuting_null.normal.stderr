error: transmuting a known null pointer into a reference
  --> tests/ui/comparison/src/transmuting_null.rs:6:23
   |
LL |         let x: &u64 = std::mem::transmute(0 as *const u64);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `0 as *const u64` is known to be a null pointer, so transmuting it into a reference is undefined behavior
   = note: `#[deny(rpl::transmuting_null)]` on by default

error: transmuting a known null pointer into a reference
  --> tests/ui/comparison/src/transmuting_null.rs:10:23
   |
LL |         let x: &u64 = std::mem::transmute(std::ptr::null::<u64>());
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `std::ptr::null::<u64>()` is known to be a null pointer, so transmuting it into a reference is undefined behavior

error: transmuting a known null pointer into a reference
  --> tests/ui/comparison/src/transmuting_null.rs:39:23
   |
LL |         let x: &u64 = std::mem::transmute(null_ptr);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `null_ptr` is known to be a null pointer, so transmuting it into a reference is undefined behavior

error: transmuting a known null pointer into a reference
  --> tests/ui/comparison/src/transmuting_null.rs:46:23
   |
LL |         let x: &u64 = std::mem::transmute(null_ptr);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `std::ptr::null::<u64>()` is known to be a null pointer, so transmuting it into a reference is undefined behavior

error: aborting due to 4 previous errors

